executor/aggregate.go

```diff
@@ -977,3 +997,258 @@ func (e *groupChecker) reset() {
 		e.tmpGroupKey = e.tmpGroupKey[:0]
 	}
 }
+
+type vecGroupChecker struct {
+	ctx                  sessionctx.Context
+	StmtCtx              *stmtctx.StatementContext
+	GroupByItems         []expression.Expression
+	groupRowsIndex       []int
+	groupRowsNum         int
+	curGroupID           int
+	previousLastGroupKey []byte
+	firstGroupKey        []byte
+	lastGroupKey         []byte
+	sameGroup            []bool
+}
+
+func newVecGroupChecker(ctx sessionctx.Context, stmtCtx *stmtctx.StatementContext, items []expression.Expression) *vecGroupChecker {
+	sameGroup := make([]bool, 1024)
+	return &vecGroupChecker{
+		ctx:          ctx,
+		StmtCtx:      stmtCtx,
+		GroupByItems: items,
+		groupRowsNum: 0,
+		curGroupID:   0,
+		sameGroup:    sameGroup,
+	}
+}
+
+// splitChunk divide the chunk into more groups which the row in the same group have the same groupKey
+// TODO: Since all the group by items are only a column reference, guaranteed by building projection below aggregation, we can directly compare data in a chunk.
+func (e *vecGroupChecker) splitChunk(chk *chunk.Chunk) (flag bool, err error) {
+	numRows := chk.NumRows()
+	e.curGroupID = 0
+	e.groupRowsIndex = e.groupRowsIndex[:0]
+	e.firstGroupKey = e.firstGroupKey[:0]
+	e.lastGroupKey = e.lastGroupKey[:0]
+	if len(e.GroupByItems) == 0 {
+		e.groupRowsIndex = append(e.groupRowsIndex, numRows)
+		return true, nil
+	}
+
+	if len(e.sameGroup) < numRows {
+		e.sameGroup = make([]bool, numRows)
+	}
+	if numRows != 0 {
+		e.sameGroup[0] = false
+	}
+	for i := 1; i < numRows; i++ {
+		e.sameGroup[i] = true
+	}
+
+	for _, item := range e.GroupByItems {
+		tp := item.GetType()
+		eType := tp.EvalType()
+		col, err := expression.GetColumn(eType, numRows)
+		if err != nil {
+			return false, err
+		}
+		err = expression.VecEval(e.ctx, item, chk, col)
+		if err != nil {
+			return false, err
+		}
+
+		e.firstGroupKey, err = codec.EncodeTo(e.StmtCtx, col, 0, e.firstGroupKey, tp)
```

Comment:
How about using `EncodeValue`?

Reply:
`EncodeValue` use `types.Datum` as its arguments which would incur performance degradation, but only two rows in every batch have this problem so it seems acceptable.